src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs

```diff
@@ -0,0 +1,386 @@
+    | x == y =
+        ( x
+        , Left
+            ( Fix $ AndPattern And
+                { andSort = variableSort x
+                , andFirst = t1
+                , andSecond = t2
+                }
+            , SubstitutionMerge x t1 t2
+            )
+        ) : normalizeSubstitution1 subst
+    | otherwise =
+        ( x, Right t1) : normalizeSubstitution1 rest
+normalizeSubstitution1 [(x, t)] = [(x, Right t)]
+normalizeSubstitution1 [] = []
+
+-- solves x = (t1 /\ t2), producing x = t /\ subst
+solveEqualityConstraint
+    :: MetadataTools level
+    -> (Variable level, CommonPurePattern level, UnificationProof level)
+    -> Either
+        (UnificationError level)
+        ( UnificationSubstitution level
+        , UnificationProof level
+        )
+solveEqualityConstraint tools (v, p, ecProof) = do
+    (solution, proof) <- simplifyAnd tools p
+    return
+        ( (v, unificationSolutionTerm solution)
+          : unificationSolutionConstraints solution
+        , CombinedUnificationProof [ecProof, proof]
+        )
+
+-- iteratively finds ocurrences of x = t1 /\ x = t2, solves them, and
+-- computes the new substitution
+normalizeSubstitution
+    :: MetadataTools level
+    -> UnificationSubstitution level
+    -> Either
+        (UnificationError level)
+        (UnificationSubstitution level, UnificationProof level)
+normalizeSubstitution tools subst =
+    case mergedEqualities of
+        [] -> return (subst, EmptyUnificationProof)
+        _ -> do
+            solvedEquals <-
+                traverse (solveEqualityConstraint tools) mergedEqualities
+            let (substs, proofs) = unzip solvedEquals
+            let combinedSubst =
+                    sortBy (compare `on` fst) (concat (remainingSubst:substs))
+            (finalSubst, proof) <- normalizeSubstitution tools combinedSubst
+            return
+                ( finalSubst
+                , CombinedUnificationProof
+                    [ CombinedUnificationProof proofs
+                    , proof
+                    ]
+                )
+  where
+    mergedSubstitution = normalizeSubstitution1 subst
+    mergedEqualities = [(x, p, pf) | (x, Left (p, pf)) <- mergedSubstitution]
+    remainingSubst = [(x, t) | (x, Right t) <- mergedSubstitution]
+
+-- ^'unificationProcedure' atempts to simplify @t1 = t2@, assuming @t1@ and @t2@
+-- are terms (functional patterns) to a substitution.
+-- If successful, it also produces a proof of how the substitution was obtained.
+-- If failing, it gives a 'UnificationError' reason for the failure.
+unificationProcedure
+    :: MetadataTools level
+    -- ^functions yielding metadata for pattern heads
+    -> CommonPurePattern level
+    -- ^left-hand-side of unification
+    -> CommonPurePattern level
+    -> Either
+        (UnificationError level)
+        (UnificationSubstitution level, UnificationProof level)
+unificationProcedure tools p1 p2
+    | p1Sort /= p2Sort
+    = Left (SortClash p1Sort p2Sort)
```

Comment:
Either decrease the indentation by two spaces, or add 4. Or move the '=' on the previous line, I'm not sure.